tests/scenario_tests/test_function_decorators.py

```diff
@@ -0,0 +1,70 @@
+from typing import Callable
+
+from slack_sdk import WebClient
+
+from slack_bolt import App, Ack, BoltResponse
+from tests.mock_web_api_server import (
+    setup_mock_web_api_server,
+    cleanup_mock_web_api_server,
+)
+from tests.utils import remove_os_env_temporarily, restore_os_env
+
+
+class NoopAck(Ack):
+    def __call__(self) -> BoltResponse:
+        pass
+
+
+class TestAppDecorators:
+    signing_secret = "secret"
+    valid_token = "xoxb-valid"
+    mock_api_server_base_url = "http://localhost:8888"
+    web_client = WebClient(token=valid_token, base_url=mock_api_server_base_url)
+
+    def setup_method(self):
+        self.old_os_env = remove_os_env_temporarily()
+        setup_mock_web_api_server(self)
+
+    def teardown_method(self):
+        cleanup_mock_web_api_server(self)
+        restore_os_env(self.old_os_env)
+
+    def test_decorators(self):
+        app = App(signing_secret=self.signing_secret, client=self.web_client)
+        ack = NoopAck()
+
+        @app.function("c")
+        def handle_function_events(body: dict):
+            assert body is not None
+
+        @handle_function_events.action("some-func-action-id")
```

Comment:
We should not go with this. Using the above method reference as a decorator object like this is not a common practice at all.

Reply:
This pattern came from the [functools library](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod) where it seems to appear in few places, The pattern seems to fit our use case well it is true that it does not appear in other web frameworks but I don't think the use case is very common

Reply:
@WilliamBergamin Ah, I didn't know these design in the standard libraries. Although I still feel this design can be confusing for some developers including me (investigating the reason why it does not work can be hard), I can agree to go with this if you have confidence on this.

Reply:
I see how details of the pattern can be missed, since I've used this pattern in the past I may be more familiar with it, I will ask other members of the team for their opinion on this 🧠 

but overall, I do feel confident with this implementation

Reply:
In this [PR](https://github.com/WilliamBergamin/bolt-python/pull/2) I've implemented a prototype that allows the use case described[ here](https://github.com/slackapi/bolt-python/pull/715#discussion_r961157041) to work without breaking the existing behavior

Reply:
Thanks, it looks great to me